### PR TITLE
Issue 117: Added support for variable messages

### PR DESCRIPTION
**Pull request by goblinhero**
# Info
This Pull Request is related to issue no. [117] ([https://github.com/valit-stack/Valit/issues/117]) 

# Changes
- Updated ExtensionMethods to support both
- Added tests to the new feature

### CODE REVIEW

**Review by dbarwikowski — DISMISSED** [review]

File: src/Valit/Errors/ValitRuleError.cs [hunk 1]
```diff
@@ -1,21 +1,25 @@
...
+        private ValitRuleError(Func<string> message)
         {
-            Message = message;
+            _message = message;
         }
 
         private ValitRuleError(int errorCode)
         {
             ErrorCode = errorCode;
+            _message = () => null;
```
Comment (GooRiOn) [inline src/Valit/Errors/ValitRuleError.cs]: I'm not quite sure whether we should return null or string.Empty as default value. @arekbal  ?

File: src/Valit/Errors/ValitRuleError.cs [hunk 2]
```diff
@@ -1,21 +1,25 @@
+using System;
+
 namespace Valit.Errors
 {
     internal class ValitRuleError
     {
-        public string Message  { get; }
+        private readonly Func<string> _message;
```
Comment (GooRiOn) [inline src/Valit/Errors/ValitRuleError.cs]: can you add "Func" sufix, just for consistency ?

File: src/Valit/ValitRulePropertyExtensions.cs [hunk 3]
```diff
@@ -44,14 +44,27 @@ public static class ValitRulePropertyExtensions
             return rule;
         }
 
+        public static IValitRule<TObject, TProperty> WithMessage<TObject, TProperty>(
+            this IValitRule<TObject, TProperty> rule, Func<string> message) where TObject : class
```
Comment (GooRiOn) [inline src/Valit/ValitRulePropertyExtensions.cs]: can you add "Func" sufix, just for consistency ?

**Review by GooRiOn — COMMENTED** [review]
I must admit, I really like this :D I took me some time to see the adventage but when I got it, it looks cool. 
If you could just change this little stylings, that would be great. Besides that the code itself LGTM.

File: src/Valit/Errors/ValitRuleError.cs (hunk 1 above)
Comment (arekbal) [reply src/Valit/Errors/ValitRuleError.cs]: I would go with ""... ;) Let it be string.Empty then.

File: src/Valit/ValitRulePropertyExtensions.cs [hunk 4]
```diff
@@ -44,14 +44,27 @@ public static class ValitRulePropertyExtensions
             return rule;
         }
 
+        public static IValitRule<TObject, TProperty> WithMessage<TObject, TProperty>(
```
Comment (arekbal) [inline src/Valit/ValitRulePropertyExtensions.cs:47]: Are you planning to add these different overloads we wrote about?

Comment (goblinhero) [reply src/Valit/ValitRulePropertyExtensions.cs:47]: I will do that in a later issue :-)

File: src/Valit/ValitRulePropertyExtensions.cs (hunk 3 above)
Comment (goblinhero) [reply src/Valit/ValitRulePropertyExtensions.cs]: It has been done :-)

File: src/Valit/Errors/ValitRuleError.cs (hunk 2 above)
Comment (goblinhero) [reply src/Valit/Errors/ValitRuleError.cs]: It has been done.

File: src/Valit/Errors/ValitRuleError.cs (hunk 1 above)
Comment (goblinhero) [reply src/Valit/Errors/ValitRuleError.cs]: I have updated the code to return string.Empty :-)

**Review by GooRiOn — APPROVED** [review]
LGTM :) Welcome aboard ;)

**Pull request merged**